test/TestPublicStealthKeyResolver.js

```diff
@@ -0,0 +1,732 @@
+            await exceptions.expectFailure(resolver.methods['setAddr(bytes32,address)'](node, accounts[0], {from: accounts[1]}));
+        });
+
+        it('permits non-owners to set authorisations', async () => {
+            await resolver.setAuthorisation(node, accounts[2], true, {from: accounts[1]});
+
+            // The authorisation should have no effect, because accounts[1] is not the owner.
+            await exceptions.expectFailure(
+                resolver.methods['setAddr(bytes32,address)'](node, accounts[0], {from: accounts[2]})
+            );
+        });
+
+        it('checks the authorisation for the current owner', async () => {
+            await resolver.setAuthorisation(node, accounts[2], true, {from: accounts[1]});
+            await ens.setOwner(node, accounts[1], {from: accounts[0]});
+
+            await resolver.methods['setAddr(bytes32,address)'](node, accounts[0], {from: accounts[2]});
+            assert.equal(await resolver.addr(node), accounts[0]);
+        });
+    });
+
+    describe('multicall', async () => {
+        it('allows setting multiple fields', async () => {
+            var addrSet = resolver.contract.methods['setAddr(bytes32,address)'](node, accounts[1]).encodeABI();
+            var textSet = resolver.contract.methods.setText(node, "url", "https://ethereum.org/").encodeABI();
+            var tx = await resolver.multicall([addrSet, textSet], {from: accounts[0]});
+
+            assert.equal(tx.logs.length, 3);
+            assert.equal(tx.logs[0].event, "AddressChanged");
+            assert.equal(tx.logs[0].args.node, node);
+            assert.equal(tx.logs[0].args.newAddress, accounts[1].toLowerCase());
+            assert.equal(tx.logs[1].event, "AddrChanged");
+            assert.equal(tx.logs[1].args.node, node);
+            assert.equal(tx.logs[1].args.a, accounts[1]);
+            assert.equal(tx.logs[2].event, "TextChanged");
+            assert.equal(tx.logs[2].args.node, node);
+            assert.equal(tx.logs[2].args.key, "url");
+
+            assert.equal(await resolver.methods['addr(bytes32)'](node), accounts[1]);
+            assert.equal(await resolver.text(node, "url"), "https://ethereum.org/");
+        });
+
+        it('allows reading multiple fields', async () => {
+            await resolver.methods['setAddr(bytes32,address)'](node, accounts[1], {from: accounts[0]});
+            await resolver.setText(node, "url", "https://ethereum.org/", {from: accounts[0]});
+            var results = await resolver.multicall.call([
+                resolver.contract.methods['addr(bytes32)'](node).encodeABI(),
+                resolver.contract.methods.text(node, "url").encodeABI()
+            ]);
+            assert.equal(web3.eth.abi.decodeParameters(['address'], results[0])[0], accounts[1]);
+            assert.equal(web3.eth.abi.decodeParameters(['string'], results[1])[0], "https://ethereum.org/");
+        });
+    });
+});
+
+function dnsName(name) {
```

Comment:
Another spot a link to the original source would be good

Reply:
Not sure what you mean here?

Reply:
Added some comments to the top of the contract and the test file 👍